cpp/include/raft/comms/std_comms.hpp

```diff
@@ -0,0 +1,406 @@
+/*
+ * Copyright (c) 2020, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <unordered_map>
+#include <unordered_set>
+#include <utility>
+
+#include <nccl.h>
+
+#include <raft/comms/comms.hpp>
+
+#include <ucp/api/ucp.h>
+#include <ucp/api/ucp_def.h>
+#include "ucp_helper.hpp"
+
+#include <nccl.h>
+
+#include <stdlib.h>
+#include <time.h>
+#include <algorithm>
+#include <chrono>
+#include <cstdio>
+#include <exception>
+#include <memory>
+#include <raft/handle.hpp>
+
+#include <thread>
+
+#include <cuda_runtime.h>
+
+#include <raft/cudart_utils.h>
+
+#define NCCL_CHECK(call)                                                       \
+  do {                                                                         \
+    ncclResult_t status = call;                                                \
+    ASSERT(ncclSuccess == status, "ERROR: NCCL call='%s'. Reason:%s\n", #call, \
+           ncclGetErrorString(status));                                        \
+  } while (0)
+
+#define NCCL_CHECK_NO_THROW(call)                         \
+  do {                                                    \
+    ncclResult_t status = call;                           \
+    if (status != ncclSuccess) {                          \
+      printf("NCCL call='%s' failed. Reason:%s\n", #call, \
+             ncclGetErrorString(status));                 \
+    }                                                     \
+  } while (0)
+
+namespace raft {
+namespace comms {
+
+size_t getDatatypeSize(const datatype_t datatype) {
+  switch (datatype) {
+    case CHAR:
+      return sizeof(char);
+    case UINT8:
+      return sizeof(uint8_t);
+    case INT:
+      return sizeof(int);
+    case UINT:
+      return sizeof(unsigned int);
+    case INT64:
+      return sizeof(int64_t);
+    case UINT64:
+      return sizeof(uint64_t);
+    case FLOAT:
+      return sizeof(float);
+    case DOUBLE:
+      return sizeof(double);
+  }
+}
+
+ncclDataType_t getNCCLDatatype(const datatype_t datatype) {
```

Comment:
Not too sure about this @afender . We are passing `datatype` at runtime (coming from the comms interface itself) and thus not sure how this would be helpful.

Reply:
BTW, for the opposite direction, however, there's already something similar (`get_type` method)

Reply:
That's fine.

Taking a step back, we should look at the motivation for the `datatype_t. We had something similar in cuGraph a few releases back for the C-API but that went away as part of C++ upgrade.

This didn't really bring limitations while allowing to better leverage native C++ types, structures, templates, and std:: functions. As @seunghwak pointed out, the emergence of device_vector to replace void buffers will further reduce the need for carrying custom enum type as well.